config/local_defaults.go

```diff
@@ -195,379 +77,198 @@ var defaultLocalV6 = Local{
+				}
+			default:
+				panic(fmt.Sprintf("unsupported data type (%s) encountered when reflecting on config.Local datatype %s", reflect.ValueOf(&defaultCurrentConfig).Elem().FieldByName(field.Name).Kind(), field.Name))
+			}
 		}
-		newCfg.Version = 2
 	}
-	// Migrate 2 -> 3
-	if newCfg.Version == 2 {
-		if newCfg.MaxConnectionsPerIP == defaultLocalV2.MaxConnectionsPerIP {
-			newCfg.MaxConnectionsPerIP = defaultLocalV3.MaxConnectionsPerIP
-		}
-		if newCfg.CatchupParallelBlocks == defaultLocalV2.CatchupParallelBlocks {
-			newCfg.CatchupParallelBlocks = defaultLocalV3.CatchupParallelBlocks
-		}
-		newCfg.Version = 3
+	return
+}
+
+func getLatestConfigVersion() uint32 {
+	localType := reflect.TypeOf(Local{})
+	versionField, found := localType.FieldByName("Version")
+	if !found {
+		return 0
 	}
-	// Migrate 3 -> 4
-	if newCfg.Version == 3 {
-		if newCfg.BroadcastConnectionsLimit == defaultLocalV3.BroadcastConnectionsLimit {
-			newCfg.BroadcastConnectionsLimit = defaultLocalV4.BroadcastConnectionsLimit
-		}
-		if newCfg.AnnounceParticipationKey == defaultLocalV3.AnnounceParticipationKey {
-			newCfg.AnnounceParticipationKey = defaultLocalV4.AnnounceParticipationKey
+	version := uint32(0)
+	for {
+		_, hasTag := reflect.StructTag(versionField.Tag).Lookup(fmt.Sprintf("version[%d]", version+1))
+		if !hasTag {
+			return version
 		}
-		if newCfg.PriorityPeers == nil {
-			newCfg.PriorityPeers = map[string]bool{}
-		}
-		newCfg.Version = 4
+		version++
 	}
-	// Migrate 4 -> 5
-	if newCfg.Version == 4 {
-		if newCfg.TxPoolSize == defaultLocalV4.TxPoolSize {
-			newCfg.TxPoolSize = defaultLocalV5.TxPoolSize
-		}
-		if newCfg.CatchupParallelBlocks == defaultLocalV4.CatchupParallelBlocks {
-			newCfg.CatchupParallelBlocks = defaultLocalV5.CatchupParallelBlocks
-		}
-		if newCfg.PeerConnectionsUpdateInterval == defaultLocalV4.PeerConnectionsUpdateInterval {
-			newCfg.PeerConnectionsUpdateInterval = defaultLocalV5.PeerConnectionsUpdateInterval
-		}
+}
 
-		newCfg.Version = 5
+func getVersionedDefaultLocalConfig(version uint32) (local Local) {
+	if version < 0 {
+		return
 	}
-
-	// Migrate 5 -> 6
-	if newCfg.Version == 5 {
-		if newCfg.DNSSecurityFlags == 0 {
-			newCfg.DNSSecurityFlags = defaultLocalV6.DNSSecurityFlags
-		}
-		if newCfg.EnablePingHandler == defaultLocalV5.EnablePingHandler {
-			newCfg.EnablePingHandler = defaultLocalV6.EnablePingHandler
-		}
-
-		newCfg.Version = 6
+	if version > 0 {
+		local = getVersionedDefaultLocalConfig(version - 1)
 	}
-
-	// Migrate 6 -> 7
-	if newCfg.Version == 6 {
-		if newCfg.CatchpointInterval == defaultLocalV6.CatchpointInterval {
-			newCfg.CatchpointInterval = defaultLocalV7.CatchpointInterval
+	// apply version specific changes.
+	localType := reflect.TypeOf(local)
+	for fieldNum := 0; fieldNum < localType.NumField(); fieldNum++ {
+		field := localType.Field(fieldNum)
+		versionDefaultValue, hasTag := reflect.StructTag(field.Tag).Lookup(fmt.Sprintf("version[%d]", version))
+		if !hasTag {
+			continue
```

Comment:
What's the intended setting of the default value for `LogArchiveName` if `version = 5` was passed in? Is it `""` or `"node.archive.log"`? I think the former is inconsistent with old behavior.

Reply:
The definition for LogArchiveName is
LogArchiveName string `version[4]:"node.archive.log"`
In the above case, calling the Lookup would return hasTag = false, which would ignore this variable for version 5.
this means that it would keep the value it had for version 4.

Reply:
Ah, right--I missed the recursive call on line 209.